pkg/hwinfo/hwinfo_nonwindows.go

```diff
@@ -22,9 +22,36 @@ func isCommandAvailable(name string) bool {
 }
 
 func fetchInventory() (map[string]interface{}, error) {
-	// check dmidecode is present in the system
+	res := make(map[string]interface{})
+
+	errorCollector := common.ErrorCollector{}
+	pciDevices := listPCIDevices(&errorCollector)
+	if len(pciDevices) > 0 {
+		res["pci.list"] = pciDevices
+	}
+
+	usbDevices := listUSBDevices(&errorCollector)
+	if len(usbDevices) > 0 {
+		res["usb.list"] = usbDevices
+	}
+
+	displays := listDisplays(&errorCollector)
+	if len(displays) > 0 {
+		res["display.list"] = displays
+	}
+
+	dmiDecodeResults := retrieveInfoUsingDmiDecode(&errorCollector)
+	if len(dmiDecodeResults) > 0 {
+		res = mergeStringMaps(&res, &dmiDecodeResults)
+	}
+
+	return res, errorCollector.Combine()
+}
+
+func retrieveInfoUsingDmiDecode(errs *common.ErrorCollector) map[string]interface{} {
```

Comment:
Can we stick to idiomatic Go function patterns like `func foo() error`. I don't like that a function related to business logic of the app has so much noise in its Inputs and Outputs. Both `*common.ErrorCollector `and `map[string]interface{}` has nothing to do with function logic, they are orchestration-scope helpers.

You can create an idiomatic function like `func foo() error` and then collect the errors in the orchestrator itself, e.g.
```go
var err error
var errs *common.ErrorCollector

err = foo()
errs.Add(err)
err = bar()
errs.Add(err)

errorCollector.Combine()
```

Also, there is a package called https://github.com/pkg/errors that is designed to provide multi-error capabilities without breaking the flow.

Reply:
I agree, that ErrorCollector looks like too much for most of the simple functions, but in complex functions like `hwinfo_other.go: listUSBDevices`, I could not find a better way to collect errors from inside of the loop. In other places, ErrorCollector is used only for consistency.

map[string]interface{} is a piece of old hwinfo implementation that I decided to leave intact in this PR, as it's not directly related to the task. In the future, we can rework it in a separate PR (for example returning meaningful types instead of string maps).

Not sure what do you mean by 'orchestration' in this context, though.
Also, not sure how `github.com/pkg/errors` can help when working with collecting errors.

Reply:
I agree, let't have this as a different task, it definitely requires a single big refactor targeting error collection & error contexts. As for 'orchestration', it named the logic that will call these functions and collect their results.

Reply:
Sorry I mistook the `github.com/pkg/errors` package with https://github.com/hashicorp/go-multierror however `pkg/errors` is cool too. They can be combined for best results.

Reply:
I agree that `pkg/errors` is neat. I use it in every of my side-projects. But I would favour an approach where we update the whole codebase at once using that package.
Should be a lot of changes but since the logic doesn't change it should be doable in a (very) short amount of time